### PR TITLE
[K/N][Tests] Fix crashes when casting to an Obj-C class companion

### PR DESCRIPTION
**Pull request by edisongz**
^KT-65260

### CODE REVIEW

File: compiler/ir/ir.objcinterop/src/org/jetbrains/kotlin/ir/objcinterop/ObjCInterop.kt [hunk 1]
```diff
@@ -266,9 +266,14 @@ fun IrClass.getExternalObjCClassBinaryName(): String =
...
                 ?: this.name.asString()
 
-fun IrClass.getExternalObjCMetaClassBinaryName(): String =
-        this.getExplicitExternalObjCClassBinaryName()
-                ?: this.name.asString().removeSuffix("Meta")
+fun IrClass.getExternalObjCMetaClassBinaryName(): String {
+    this.getExplicitExternalObjCClassBinaryName()?.let { return it }
+    // External ObjC metaclass is named as `Foo.Componion`, Foo is subclass of NSObject or itself.
+    if (this.name.asString() == "Companion") {
+        this.symbol.signature?.asPublic()?.declarationFqName?.let { return it.removeSuffix(".Companion") }
```
Comment (SvyatoslavScherbina) [inline compiler/ir/ir.objcinterop/src/org/jetbrains/kotlin/ir/objcinterop/ObjCInterop.kt:273]: This seems a bit fragile. Please use `.isCompanion` and `.parent` for a more robust solution.

**Review by SvyatoslavScherbina — CHANGES_REQUESTED** [review]
Please add tests covering the change.

File: kotlin-native/backend.native/tests/interop/objc/kt65260.kt [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+import platform.Foundation.NSArray
+import platform.Foundation.NSString
+import platform.darwin.NSObject
+
+fun main() {
+    // Test: cast to NSObject Metaclass
+    Any() as NSObject.Companion
+    // Test: cast to subclass of NSObject Metaclass
+    Any() as NSString.Companion
+    Any() as NSArray.Companion
```
Comment (SvyatoslavScherbina) [inline kotlin-native/backend.native/tests/interop/objc/kt65260.kt:15]: All these casts should fail by definition. Do they?

For example, `NSObject.Companion` is a companion object class, meaning that the only instance of this class is the companion object. In other words, `NSObject as NSObject.Companion` should succeed, but `Any() as NSObject.Companion` should fail (i.e. throw an exception).

File: compiler/ir/ir.objcinterop/src/org/jetbrains/kotlin/ir/objcinterop/ObjCInterop.kt [hunk 3]
```diff
@@ -266,9 +267,14 @@ fun IrClass.getExternalObjCClassBinaryName(): String =
...
                 ?: this.name.asString()
 
-fun IrClass.getExternalObjCMetaClassBinaryName(): String =
-        this.getExplicitExternalObjCClassBinaryName()
-                ?: this.name.asString().removeSuffix("Meta")
+fun IrClass.getExternalObjCMetaClassBinaryName(): String {
+    this.getExplicitExternalObjCClassBinaryName()?.let { return it }
+    // External ObjC metaclass is named as `Foo.Componion`, Foo is subclass of NSObject or itself.
+    if (this.isCompanion) {
+        (this.parent as? IrLazyClass)?.let { return it.name.asString() }
```
Comment (SvyatoslavScherbina) [inline compiler/ir/ir.objcinterop/src/org/jetbrains/kotlin/ir/objcinterop/ObjCInterop.kt:274]: `IrLazyClass` is an implementation detail. It should be enough to cast to `IrClass` instead.

File: kotlin-native/backend.native/tests/build.gradle [hunk 4]
```diff
@@ -1103,6 +1103,11 @@ if (PlatformInfo.isAppleTarget(project)) {
         UtilsKt.dependsOnPlatformLibs(it)
     }
 
+    standaloneTest("interop_objc_kt65260") {
+        source = "interop/objc/kt65260.kt"
+        UtilsKt.dependsOnPlatformLibs(it)
+    }
```
Comment (SvyatoslavScherbina) [inline kotlin-native/backend.native/tests/build.gradle:1109]: This file is the legacy test infrastructure. We are currently migrating all tests from here.
Please add the new test there instead: https://github.com/JetBrains/kotlin/tree/master/native/native.tests/testData/codegen/cinterop/objc
After creating a file, you will need to run `generateTests` Gradle task (or `Generate Tests` run configuration in IDEA). This will generate JUnit tests for the new testdata file.

**Review by SvyatoslavScherbina — CHANGES_REQUESTED** [review]

File: compiler/ir/ir.objcinterop/src/org/jetbrains/kotlin/ir/objcinterop/ObjCInterop.kt (hunk 3 above)
Comment (edisongz) [reply compiler/ir/ir.objcinterop/src/org/jetbrains/kotlin/ir/objcinterop/ObjCInterop.kt]: Fixed

File: kotlin-native/backend.native/tests/build.gradle (hunk 4 above)
Comment (edisongz) [reply kotlin-native/backend.native/tests/build.gradle]: Tests have been migrated.

File: kotlin-native/backend.native/tests/interop/objc/kt65260.kt (hunk 2 above)
Comment (edisongz) [reply kotlin-native/backend.native/tests/interop/objc/kt65260.kt]: Yes, all these tests will throw `kotlin.TypeCastException` in runtime.
```
Any() as NSObject.Companion

Uncaught Kotlin exception: kotlin.TypeCastException: class kotlin.Any cannot be cast to class platform.darwin.NSObject.Companion
```

`NSObject as NSObject.Companion` compile works, but it‘s unnecessary, also IDEA prompts `No cast needed` warning.
```kotlin
// Two expressions are equivalent
val nsObject = NSObject
val nsObjectComponion = NSObject.Companion
```

File: compiler/ir/ir.objcinterop/src/org/jetbrains/kotlin/ir/objcinterop/ObjCInterop.kt [hunk 5]
```diff
@@ -266,9 +266,14 @@ fun IrClass.getExternalObjCClassBinaryName(): String =
...
                 ?: this.name.asString()
 
-fun IrClass.getExternalObjCMetaClassBinaryName(): String =
-        this.getExplicitExternalObjCClassBinaryName()
-                ?: this.name.asString().removeSuffix("Meta")
+fun IrClass.getExternalObjCMetaClassBinaryName(): String {
+    this.getExplicitExternalObjCClassBinaryName()?.let { return it }
+    // External ObjC metaclass is named as `Foo.Componion`, Foo is subclass of NSObject or itself.
+    if (this.isCompanion) {
+        (this.parent as? IrClass)?.let { return it.name.asString() }
```
Comment (SvyatoslavScherbina) [inline compiler/ir/ir.objcinterop/src/org/jetbrains/kotlin/ir/objcinterop/ObjCInterop.kt:273]: What if the class has an `@ExternalObjCClass(binaryName = )` annotation? Shouldn't we return that value then, as in other cases?

Please add a test for this.

File: native/native.tests/testData/codegen/cinterop/objc/kt65260.kt [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+import kotlin.test.*
+
+fun box(): String {
+    try {
+        // Test: cast to NSObject Metaclass
+        Any() as NSObject.Companion
+        // Test: cast to subclass of NSObject Metaclass
+        Any() as NSString.Companion
+        Any() as NSArray.Companion
+    } catch (e: Exception) {
```
Comment (SvyatoslavScherbina) [inline native/native.tests/testData/codegen/cinterop/objc/kt65260.kt:19]: This test, in fact, checks that at least one of these casts fails. But it really should check that all of them fail.

Also, it makes sense to verify `is` checks as well.

File: kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/CodeGenerator.kt [hunk 7]
```diff
@@ -1331,7 +1331,7 @@ internal abstract class FunctionGenerationContext(
             generationState.dependenciesTracker.add(irClass)
             if (irClass.isObjCMetaClass()) {
                 val name = irClass.getExternalObjCMetaClassBinaryName()
-                val objCClass = getObjCClass(name)
+                val objCClass = getObjCClass(name, isMetaclass = true)
```
Comment (SvyatoslavScherbina) [inline kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/CodeGenerator.kt:1334]: This change doesn't seem right: now the following code
```kotlin
val nsObjectClass: Any = NSObject
println(nsObjectClass is NSNumberMeta)
```
prints `true`, which is not correct. It should print `false`, as it did before. Please add such a test case.

File: kotlin-native/backend.native/tests/interop/objc/kt65260.kt (hunk 2 above)
Comment (SvyatoslavScherbina) [reply kotlin-native/backend.native/tests/interop/objc/kt65260.kt]: > `NSObject as NSObject.Companion` compile works, but it‘s unnecessary, also IDEA prompts `No cast needed` warning.

Checking this is still necessary, because the compiler can't always rely on compile-time types. For example,
```kotlin
val nsObjectClass: Any = NSObject
nsObjectClass as NSObject.Companion
```
In this case, the compiler will generate the actual cast.

**Review by SvyatoslavScherbina — CHANGES_REQUESTED** [review]
The behavior still doesn't seem entirely right:
```kotlin
NSNumber as NSObject.Companion
```
such a cast should fail, but it succeeds.

**Comment by edisongz** [comment]
> The behavior still doesn't seem entirely right:
> 
> ```kotlin
> NSNumber as NSObject.Companion
> ```
> 
> such a cast should fail, but it succeeds.

I have a doubt about this case, corresponding to ObjC.
```objectivec
// This result is YES, ObjC class object is instance of meta class.
BOOL result = [NSNumber isKindOfClass:object_getClass(NSObject.class)]; 
```

So `ThrowTypeCastException` branch cannot be reached.
image

File: kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/CodeGenerator.kt (hunk 7 above)
Comment (edisongz) [reply kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/CodeGenerator.kt]: `object_getClass` returns meta class, These pieces of code has been reverted.

File: compiler/ir/ir.objcinterop/src/org/jetbrains/kotlin/ir/objcinterop/ObjCInterop.kt (hunk 5 above)
Comment (edisongz) [reply compiler/ir/ir.objcinterop/src/org/jetbrains/kotlin/ir/objcinterop/ObjCInterop.kt:273]: If the class has an`@ExternalObjCClass(binaryName = )` annotation,  it returns the binary name immediately.
The purpose of the `isCompanion` code block is also to obtain the class name.

As for test cases, it seems that this usage is not supported yet; I searched the entire `kotlin` repository and found no usage. Are there any examples?

`java.lang.IllegalArgumentException: External objc Classes can't be used this way`

```kotlin
@OptIn(BetaInteropApi::class)
@ExternalObjCClass(binaryName = "Foo")
open class A: NSObject() {
    companion object
}

@OptIn(BetaInteropApi::class)
@ExternalObjCClass(binaryName = "Bar")
open class B: NSObject() {
    companion object
}

B as A.Companion
```

Comment (SvyatoslavScherbina) [reply compiler/ir/ir.objcinterop/src/org/jetbrains/kotlin/ir/objcinterop/ObjCInterop.kt:273]: > As for test cases, it seems that this usage is not supported yet; I searched the entire kotlin repository and found no usage. Are there any examples?

`@ExternalObjCClass(binaryName = )` is generated by `cinterop` for classes with `objc_runtime_name` attribute. You are not supposed to add this annotation manually to Kotlin code.

>  If the class has an`@ExternalObjCClass(binaryName = )` annotation, it returns the binary name immediately,
The purpose of the `isCompanion` code block is also to obtain the class name.

If an Objective-C class has `objc_runtime_name` attribute, then `getExternalObjCMetaClassBinaryName` should return this name.

**Comment by SvyatoslavScherbina** [comment]
> > The behavior still doesn't seem entirely right:
> > ```kotlin
> > NSNumber as NSObject.Companion
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > such a cast should fail, but it succeeds.
> 
> I have a doubt about this case, corresponding to ObjC.
> 
> ```objc
> // This result is YES, ObjC class object is instance of meta class.
> BOOL result = [NSNumber isKindOfClass:object_getClass(NSObject.class)]; 
> ```

Yes, but semantically `NSNumber as NSObject.Companion` is not the same.
`[NSNumber isKindOfClass:object_getClass(NSObject.class)]` is equivalent to `NSNumber is NSObjectMeta`, which should be `true`.
But `NSNumber is NSObject.Companion` has no equivalent in Objective-C. `NSObject.Companion` is a class that Kotlin synthesizes to reflect the concept of the class object.


`NSNumber is NSNumber.Companion == true` contradicts general Kotlin rules.
Consider e.g. the following example:
```kotlin
open class A {
    companion object : AMeta()
}
open class AMeta

open class B : A() {
    companion object : BMeta()
}
open class BMeta : AMeta()

fun main() {
    val bAny: Any = B
    println(bAny is A.Companion)
}
```

It prints `false`.
`A.Companion` is a class with only one instance -- `A`. So `B` can't be an instance of `A.Companion`.
Same with `NSObject.Companion` and `NSNumber`: `NSNumber` cant be an instance of `NSObject.Companion`, only `NSObject` is.

File: kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt [hunk 8]
```diff
@@ -1667,6 +1672,16 @@ internal class CodeGeneratorVisitor(
...
 
+    private fun isCompanionParentClass(value: IrTypeOperatorCall, dstClass: IrClass): Boolean {
+        var result = true
+        if (dstClass.isCompanion) {
+            (dstClass.parent as? IrClass)?.symbol?.let {
+                result = value.argument.type == it
+            }
+        }
+        return result
+    }
```
Comment (SvyatoslavScherbina) [inline kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt:1683]: I don't really understand what was the intention, but the result doesn't seem correct:
```kotlin
class Foo {
    companion object
}

fun main() {
    val foo: Any = Foo
    println(foo is Foo.Companion)
}
```
prints `false` now, which is wrong.

**Review by SvyatoslavScherbina — CHANGES_REQUESTED** [review]

File: kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt (hunk 8 above)
Comment (edisongz) [reply kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt:1683]: Wrong `if` judgment, only covering specific `NSObject` test cases.

File: kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt [hunk 9]
```diff
@@ -1686,7 +1686,12 @@ internal class CodeGeneratorVisitor(
                 null
         )
 
-        return if (dstClass.isObjCClass()) {
+        return if (dstClass.isCompanion) {
+            call(
+                    llvm.Kotlin_Interop_IsSameClass,
+                    listOf(objCObject, genGetObjCClass(dstClass))
+            )
```
[the remark this answers is not in the document]
Comment (edisongz) [reply kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt:1693]: If use pointer equality for type-checking, the above testcases print.
```kotlin
val nsObjectClass: ObjCClass = NSObject.Companion
// false
println(nsObjectClass is NSObject.Companion)

val nsObjectCompanion = NSObject.Companion
// true
println(nsObjectCompanion is NSObject.Companion)

// true
println(nsObjectMetaClass is NSObject.Companion)

// false
println(nsObjectClass == nsObjectMetaClass)
```

For me, the first case is confusing, it's so different with `C-family` languages or `Swift`.

File: compiler/ir/ir.objcinterop/src/org/jetbrains/kotlin/ir/objcinterop/ObjCInterop.kt [hunk 10]
```diff
@@ -266,9 +266,16 @@ fun IrClass.getExternalObjCClassBinaryName(): String =
...
+        val parent = this.parent as? IrClass
+        parent?.getExplicitExternalObjCClassBinaryName()?.let { return it }
+        parent?.let { return it.name.asString() }
+    }
+    return this.name.asString().removeSuffix("Meta")
+}
 
 private fun IrClass.getExplicitExternalObjCClassBinaryName() =
-        this.annotations.findAnnotation(externalObjCClassFqName)!!.getAnnotationValueOrNull<String>("binaryName")
+        this.annotations.findAnnotation(externalObjCClassFqName)?.getAnnotationValueOrNull<String>("binaryName")
```
[the remark this answers is not in the document]
Comment (edisongz) [reply compiler/ir/ir.objcinterop/src/org/jetbrains/kotlin/ir/objcinterop/ObjCInterop.kt]: Reverted.

File: kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (edisongz) [reply kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt]: `functionGenerationContext.icmpEq(objCObject, genGetObjCClass(dstClass.parentAsClass))` is right.

File: compiler/ir/ir.objcinterop/src/org/jetbrains/kotlin/ir/objcinterop/ObjCInterop.kt [hunk 11]
```diff
@@ -266,9 +266,16 @@ fun IrClass.getExternalObjCClassBinaryName(): String =
...
-fun IrClass.getExternalObjCMetaClassBinaryName(): String =
-        this.getExplicitExternalObjCClassBinaryName()
-                ?: this.name.asString().removeSuffix("Meta")
+fun IrClass.getExternalObjCMetaClassBinaryName(): String {
+    this.getExplicitExternalObjCClassBinaryName()?.let { return it }
+    // External ObjC metaclass is named as `Foo.Componion`, Foo is subclass of NSObject or itself.
+    if (this.isCompanion) {
+        val parent = this.parent as? IrClass
+        parent?.getExplicitExternalObjCClassBinaryName()?.let { return it }
+        parent?.let { return it.name.asString() }
```
[the remark this answers is not in the document]
Comment (edisongz) [reply compiler/ir/ir.objcinterop/src/org/jetbrains/kotlin/ir/objcinterop/ObjCInterop.kt]: Fixed

File: kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt [hunk 12]
```diff
@@ -1686,7 +1686,12 @@ internal class CodeGeneratorVisitor(
                 null
         )
 
-        return if (dstClass.isObjCClass()) {
+        return if (dstClass.isCompanion) {
+            call(
+                    llvm.Kotlin_Interop_IsObjectSameClass,
```
[the remark this answers is not in the document]
Comment (edisongz) [reply kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt]: Fixed.

File: kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (edisongz) [reply kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt]: > That doesn't seem entirely right.
> 
> ```kotlin
> import kotlinx.cinterop.*
> import platform.darwin.*
> import platform.objc.*
> 
> @OptIn(kotlinx.cinterop.BetaInteropApi::class)
> fun main() {
>     val nsObjectClass: ObjCClass = NSObject
>     val nsObjectMetaClass: ObjCClass = object_getClass(nsObjectClass)!!
>     println(nsObjectClass is NSObject.Companion)
>     println(nsObjectMetaClass is NSObject.Companion)
>     println(nsObjectClass == nsObjectMetaClass)
> }
> ```
> 
> For me it prints
> 
> ```
> true
> true
> false
> ```
> 
> Meaning that there are two different objects, but both are instances of `NSObject.Companion`.
> 
> It might be more straightforward and correct to compare pointer equality when type-checking against such a companion class. What do you think?

`functionGenerationContext.icmpEq(objCObject, genGetObjCClass(dstClass.parentAsClass))` is the right thing.
The above testcase will print
```
val nsObjectClass: ObjCClass = NSObject
val nsObjectMetaClass: ObjCClass = object_getClass(nsObjectClass)!!

// true
println(nsObjectClass is NSObject.Companion)

// false
println(nsObjectMetaClass is NSObject.Companion)

// false
println(nsObjectClass == nsObjectMetaClass)
```

File: compiler/ir/ir.objcinterop/src/org/jetbrains/kotlin/ir/objcinterop/ObjCInterop.kt [hunk 13]
```diff
@@ -266,9 +266,12 @@ fun IrClass.getExternalObjCClassBinaryName(): String =
         this.getExplicitExternalObjCClassBinaryName()
                 ?: this.name.asString()
 
-fun IrClass.getExternalObjCMetaClassBinaryName(): String =
-        this.getExplicitExternalObjCClassBinaryName()
-                ?: this.name.asString().removeSuffix("Meta")
+fun IrClass.getExternalObjCMetaClassBinaryName(): String {
+    this.getExplicitExternalObjCClassBinaryName()?.let { return it }
+    // External ObjC metaclass is named as `Foo.Componion`, Foo is subclass of NSObject or itself.
+    if (this.isCompanion) return (this.parent as IrClass).getExternalObjCClassBinaryName()
```
Comment (SvyatoslavScherbina) [inline compiler/ir/ir.objcinterop/src/org/jetbrains/kotlin/ir/objcinterop/ObjCInterop.kt:272]: This addition probably is not used anymore.
For example, the function is no longer used for a `companion` when generating a type check. Also, if it is used for a `companion` somewhere, it will just crash, because `getExplicitExternalObjCClassBinaryName` will throw NPE, as before.

Please revert the change.

File: native/native.tests/testData/codegen/cinterop/objc/kt65260.kt [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+class BNativeHeir : B() {
+    companion object
+}
+
+fun testExternalObjCMetaClassCast() {
+    try {
+        BNativeHeir as A.Companion
+    } catch (e: Exception) {
+        assertTrue(e is ClassCastException)
+    }
```
[the remark this answers is not in the document]
Comment (SvyatoslavScherbina) [reply native/native.tests/testData/codegen/cinterop/objc/kt65260.kt:59]: The purpose of a test is to anticipate and prevent bugs. If the test is inteded to check that the cast fails, then it does that wrong, see my first comment.

In other words, consider there is a bug (e.g. a regression) in the compiler, and the cast no longer fails. Well, the test won't report this then.

Please either use `assertFailsWith`, or check `BNativeHeir is A.Companion` instead. 

Same in many places below.

File: native/native.tests/testData/codegen/cinterop/objc/kt65260.kt [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+    assertTrue(NSData is NSData.Companion)
+}
+
+fun testNonObjCObjectTypeCast() {
+    val foo: Any = FooK
+    assertTrue(foo is FooK.Companion)
+
+    try {
+        BarK as FooK.Companion
```
Comment (SvyatoslavScherbina) [inline native/native.tests/testData/codegen/cinterop/objc/kt65260.kt:127]: This function seems inconsistent: it uses `val foo: Any = ` for `FooK`, but not for `BarK`.

This trick is required to fool the compiler, to make sure it doesn't optimize the cast early, so that the test actually checks the type check code generation. The trick is useful in both checks, above and below.

Same in other places.

**Review by SvyatoslavScherbina — CHANGES_REQUESTED** [review]

File: native/native.tests/testData/codegen/cinterop/objc/kt65260.kt (hunk 14 above)
[the remark this answers is not in the document]
Comment (edisongz) [reply native/native.tests/testData/codegen/cinterop/objc/kt65260.kt]: Use `assertFailsWith` instead.

**Comment by SvyatoslavScherbina** [comment]
Thank you for the contribution!
Merged manually as 450bb1c7c0f90c2a07ceb8208f0f23268decedf0.
